automation/Extension/firefox/lib/http-post-parser.js

```diff
@@ -0,0 +1,250 @@
+// Incorporates code from: https://github.com/redline13/selenium-jmeter/blob/6966d4b326cd78261e31e6e317076569051cac37/content/library/recorder/HttpPostParser.js
+
+const {Cc, Ci, CC, Cu, components} = require("chrome");
+var loggingDB     = require("./loggingdb.js");
+
+var HttpPostParser = function(stream) {
+  // Scriptable Stream Constants
+  this.seekablestream = stream;
+  this.stream = components.classes["@mozilla.org/scriptableinputstream;1"].createInstance(components.interfaces.nsIScriptableInputStream);
+  this.stream.init(this.seekablestream);
+
+  this.postBody = "";
+  this.postLines = [];
+  this.postHeaders = [];
+  // Check if the stream has headers
+  this.hasheaders = false;
+  this.body = 0;
+  if (this.seekablestream instanceof components.interfaces.nsIMIMEInputStream) {
+    this.seekablestream.QueryInterface(components.interfaces.nsIMIMEInputStream);
+    this.hasheaders = true;
+    this.body = -1;
+  } else if (this.seekablestream instanceof components.interfaces.nsIStringInputStream) {
+    this.seekablestream.QueryInterface(components.interfaces.nsIStringInputStream);
+    this.hasheaders = true;
+    this.body = -1;
+  }
```

Comment:
See my new comments in the updated code.

These two stream types may contain headers which is verified by checking the stream types and upstream headers encountered in our `TestPOSTInstrument` tests.

During the verification I found that `nsIMultiplexInputStream` (used when the AJAX payload is `FormData`) may also contain headers and added a check for it.

MDN lists tons of other stream types, although I think not all of them can be used in the POST context:
https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Guide/Streams

We don't really know if these other stream types may contain headers or not (not documented).

We added a debug log for unknown stream types, which we can investigate later.